aptos-move/framework/src/natives/cryptography/multi_ed25519.rs

```diff
@@ -70,6 +73,35 @@ fn native_public_key_validate(
     Ok(NativeResult::ok(cost, smallvec![Value::bool(all_valid)]))
 }
 
+/// See `public_key_validate_v2_internal` comments in `multi_ed25519.move`.
+fn native_public_key_validate_v2(
+    gas_params: &GasParameters,
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 1);
+
+    let pks_bytes = pop_arg!(arguments, Vec<u8>);
+
+    let mut cost = gas_params.base;
+
+    let mut num_deserializations = 0;
+    let mut num_small_order_checks = 0;
+
+    let all_valid = multi_ed25519::MultiEd25519PublicKey::validate_bytes_and_count_checks(
+        pks_bytes.as_slice(),
+        &mut num_deserializations,
+        &mut num_small_order_checks,
+    );
+
+    cost += gas_params.per_pubkey_deserialize * NumArgs::new(num_deserializations as u64)
+        + gas_params.per_pubkey_small_order_check * NumArgs::new(num_small_order_checks as u64);
```

Comment:
I assume it is for saving gas? It also creates the unfairness - only because the bad sub-PK comes last I'm paying more for the validation...

Also I'm not sure how important it is to save gas, but IMO it is better for a validation primitive to behave as consistent as possible (in terms of gas charge/time consumed) regardless of the validation result.

Thoughts?

Reply:
I agree but for another reason, this is an over optimization on a case that really shouldn't happen all that often. If a user simulates their txn, it would almost always discover this anyway. So the only likely cases where you hit this are when you just have something wrong with your setup and you aren't using simulation or transient issue such as someone is performing some sort of front running attack to just charge you gas

Reply:
Both of these are valid objections.

Unless you feel very strongly to not include this optimization, I plan to keep it.